README.md
C Smart Pointers
================

[![Build Status](https://travis-ci.org/Snaipe/c-smart-pointers.svg?branch=master)](https://travis-ci.org/Snaipe/c-smart-pointers) 
[![Coverage Status](https://coveralls.io/repos/Snaipe/c-smart-pointers/badge.svg?branch=master)](https://coveralls.io/r/Snaipe/c-smart-pointers?branch=master) 
[![License](https://img.shields.io/badge/license-MIT-blue.svg?style=flat)](https://github.com/Snaipe/c-smart-pointers/blob/master/LICENSE) 
[![Version](https://img.shields.io/github/tag/Snaipe/c-smart-pointers.svg?label=version&style=flat)](https://github.com/Snaipe/c-smart-pointers/releases)

## What this is

This project is a tentative attempt to bring smart pointer constructs
to the (GNU) C programming language.

### Features

* `unique_ptr`, `shared_ptr` macros, and `smart` type attribute
* Destructor support for cleanup
* Custom variable metadata on allocation
* Cross-platform: tested under linux 3.18.6-1, Mac OS X Yosemite 10.10, and Windows 7 (with MinGW and the Cygwin port of GCC)

## Installing

### With a package manager

* Mac OS X: `brew install snaipe/soft/libcsptr`

### Building from source
#### Prerequisites

To compile the library, GCC 4.9+ is needed.

#### Installation

1. Clone this repository
2. run `./autogen.sh && ./configure --prefix=$HOME/ && make && make install`
   from the project root for a local install, or run
   `./autogen.sh && ./configure && make && sudo make install` for a global
   install.

## Examples

* Simple unique\_ptr:
    simple1.c:
    ```c
    #include <stdio.h>
    #include <csptr/smart_ptr.h>

    int main(void) {
        // some_int is an unique_ptr to an int with a value of 1.
        smart int *some_int = unique_ptr(int, 1);

        printf("%p = %d\n", some_int, *some_int);

        // some_int is destroyed here
        return 0;
    }
    ```
    Shell session:
    ```bash
    $ gcc -std=c99 -o simple1 simple1.c -lcsptr
    $ valgrind ./simple1
    ==3407== Memcheck, a memory error detector
    ==3407== Copyright (C) 2002-2013, and GNU GPL\'d, by Julian Seward et al.
    ==3407== Using Valgrind-3.10.0 and LibVEX; rerun with -h for copyright info
    ==3407== Command: ./test1
    ==3407==
    0x53db068 = 1
    ==3407==
    ==3407== HEAP SUMMARY:
    ==3407==     in use at exit: 0 bytes in 0 blocks
    ==3407==   total heap usage: 1 allocs, 1 frees, 48 bytes allocated
    ==3407==
    ==3407== All heap blocks were freed -- no leaks are possible
    ==3407==
    ==3407== For counts of detected and suppressed errors, rerun with: -v
    ==3407== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
    ```
* Simple unique\_ptr with destructor:
    ```c
    #include <unistd.h>
    #include <fcntl.h>
    #include <csptr/smart_ptr.h>

    struct log_file {
        int fd;
        // ...
    };

    void cleanup_log_file(void *ptr, void *meta) {
        (void) meta;
        close(((struct log_file *) ptr)->fd);
    }

    int main(void) {
        smart struct log_file *log = unique_ptr(struct log_file, {
                .fd = open("/dev/null", O_WRONLY | O_APPEND),
                // ...
            }, cleanup_log_file);

        write(log->fd, "Hello", 5);

        // cleanup_log_file is called, then log is freed
        return 0;
    }
    ```
* Allocating a smart array and printing its contents before destruction:
    ```c
    #include <stdio.h>
    #include <csptr/smart_ptr.h>
    #include <csptr/array.h>

    void print_int(void *ptr, void *meta) {
        (void) meta;
        // ptr points to the current element
        // meta points to the array metadata (global to the array), if any.
        printf("%d\n", *(int*) ptr);
    }

    int main(void) {
        // Destructors for array types are run on every element of the
        // array before destruction.
        smart int *ints = unique_ptr(int[5], {5, 4, 3, 2, 1}, print_int);
        // ints == {5, 4, 3, 2, 1}

        // Smart arrays are length-aware
        for (size_t i = 0; i < array_length(ints); ++i) {
            ints[i] = i + 1;
        }
        // ints == {1, 2, 3, 4, 5}

        return 0;
    }
    ```

## More examples

* Using a different memory allocator (although most will replace malloc/free):
    ```c
    #include <csptr/smart_ptr.h>

    void *some_allocator(size_t);
    void some_deallocator(void *);

    int main(void) {
        smalloc_allocator = (s_allocator) {some_allocator, some_deallocator};
        // ...
        return 0;
    }
    ```

* Automatic cleanup on error cases:
    ```c
    #include <unistd.h>
    #include <fcntl.h>
    #include <csptr/smart_ptr.h>

    struct log_file {
        int fd;
        // ...
    };

    static void close_log(void *ptr, void *meta) {
        (void) meta;
        struct log_file *log = ptr;
        if (log->fd != -1)
            close(log->fd);
    }

    struct log_file *open_log(const char *path) {
        smart struct log_file *log = shared_ptr(struct log_file, {0}, close_log);
        if (!log) // failure to allocate
            return NULL; // nothing happens, destructor is not called

        log->fd = open(path, O_WRONLY | O_APPEND | O_CREAT, 0644);
        if (log->fd == -1) // failure to open
            return NULL; // log gets destroyed, file descriptor is not closed since fd == -1.

        return sref(log); // a new reference on log is returned, it does not get destoyed
    }

    int main(void) {
        smart struct log_file *log = open_log("/dev/null");
        // ...
        return 0; // file descriptor is closed, log is freed
    }
    ```
* Using named parameters:
    ```c
    #include <csptr/smart_ptr.h>

    void nothing(void *ptr, void *meta) {}

    int main(void) {
        struct { int a; } m = { 1 };

        smart int *i = unique_ptr(int,
                .dtor = nothing,
                .value = 42,
                .meta = { &m, sizeof (m) }
            );

        return 0;
    }
    ```

## FAQ

**Q. Why didn't you use C++ you moron ?**  
A. Because when I first started this, I was working on a C project.
   Also, because it's fun.

**Q. Can I use this on a serious project ?**  
A. Yes, but as this project has not been widely used, there might be
   some bugs. Beware!

**Q. How did you make this ?**  
A. Here's a [link to my blog post](http://snaipe.me/c/c-smart-pointers/) on the matter.


ChangeLog
2015-03-18  Franklin "Snaipe" Mathieu  <franklinmathieu@gmail.com>

    * include/csptr/{smart_ptr.h,apply.h}: Removed the destructor helper macro
    * include/csptr/{smart_ptr.h,smalloc.h}: Changed the vararg usage to
    struct vararg. This change is **NOT** backward-compatible. Parenthesis
    around the value were removed with this change, and metadata is no longer
    passed as two separate parameters.

2015-01-26  Franklin "Snaipe" Mathieu  <franklinmathieu@gmail.com>

    * include/csptr/smart_ptr.h: Smart pointers now take a mandatory value.
    * doc/csptr.3: Fixed some typos and updated to latest interface.

2015-01-20  Franklin "Snaipe" Mathieu  <franklinmathieu@gmail.com>

    * configure.ac: Removed --with-malloc-wrapper ./configure flag.
    * src/wrap_malloc.{c,h}: Removed malloc/free wrappers.
    * check/*: Added unit tests & coverage reports.
    * mman.c: Fix for shared pointers accessing the wrong metadata.

2015-01-14  Franklin "Snaipe" Mathieu  <franklinmathieu@gmail.com>

    * configure.ac: Added --with-malloc-wrapper and --with-fixed-allocator
    flags for ./configure.

    * src/mman.c: Added a fixed allocator support -- if the user does not
    want to override at runtime the allocators used by smalloc/sfree and just
    want to use malloc and free, they may enable this option using the
    --with-fixed-allocator configuration flag.

    * src/wrap_malloc.{c,h}, src/mman.c: Added malloc/free wrappers.
    Pointers produced by malloc can be used with the `smart` attribute,
    and smart pointers may be passed to free. This option is deactivated
    by default, and can be enabled using the --with-malloc-wrapper
    configuration flag.

2015-01-14  Franklin "Snaipe" Mathieu  <franklinmathieu@gmail.com>

    * src/smart_ptr.h: unique_ptr and shared_ptr now take a type parameter
    instead of a size.

    * src/smart_ptr.h: Added support for static array types in the unique_ptr
    and shared_ptr macros.

    * src/mman.{c,h}: Added smart array allocation through smalloc. smalloc
    has been changed to take a mandatory `nmemb` parameter, that may be 0
    if the allocated type is a scalar, and an array for any other value.

2015-01-13  Franklin "Snaipe" Mathieu  <franklinmathieu@gmail.com>

    * src/smart_ptr.h: Added unique_ptr and shared_ptr macros, added
    `smart` variable attribute.

    * src/mman.{c,h}: Added smalloc, sfree, sref, and get_smart_user_meta.
    Alternative allocators may be specified by setting the contents of
    `smalloc_allocator`.


src/vararg.h
/*
 * The MIT License (MIT)
 *
 * Copyright © 2015 Franklin "Snaipe" Mathieu <http://snai.pe/>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#ifndef CSPTR_CPP_VARARG_H_
# define CSPTR_CPP_VARARG_H_

#if __STRICT_ANSI__
# define ARG_LENGTH(...) __builtin_choose_expr(sizeof (#__VA_ARGS__) == 1,  \
        0,                                                                  \
        ARG_LENGTH__(__VA_ARGS__))
#else /* !__STRICT_ANSI__ */
# define ARG_LENGTH(...) ARG_LENGTH__(__VA_ARGS__)
#endif /* !__STRICT_ANSI__ */

# define ARG_LENGTH__(...) ARG_LENGTH_(,##__VA_ARGS__,                         \
    63, 62, 61, 60, 59, 58, 57, 56, 55, 54, 53, 52, 51, 50, 49, 48, 47, 46, 45,\
    44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26,\
    25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6,\
    5, 4, 3, 2, 1, 0)
# define ARG_LENGTH_(_, _63, _62, _61, _60, _59, _58, _57, _56, _55, _54, _53, \
    _52, _51, _50, _49, _48, _47, _46, _45, _44, _43, _42, _41, _40, _39, _38, \
    _37, _36, _35, _34, _33, _32, _31, _30, _29, _28, _27, _26, _25, _24, _23, \
    _22, _21, _20, _19, _18, _17, _16, _15, _14, _13, _12, _11, _10, _9, _8,   \
    _7, _6, _5, _4, _3, _2, _1, Count, ...) Count

#endif /* !CSPTR_CPP_VARARG_H_ */
